add poisson regression example

joelgrus opened the pull request:
it's an almost verbatim translation of the C++ example, also required bindings for the `Dict` class from `dict.h`.

I figured that implementing the examples would help me understand what the most useful parts to wrap were.

two caveats:
(1) I don't *understand* the example (it's not documented in the slightest), but the Scala version *seems* to be doing the right thing (its errors get really small)

(2) I ran into a hairy issue where `LSTMBuilder::back()` returns an `Expression` struct *by value*. It turns out that SWIG does not like returning structs by value. I tried a lot of pointer-y things to make it work, and none of them were successful.

Eventually I gave up and wrote functions that returned the two (non-class) elements of the struct, and I used Scala implicits to add a `back()` method to the Scala version that does what the C++ method would have done. 

Let me know if you hate this approach and/or if you have an idea how to make it not necessary.

jayantk commented:
(I know we just talked about it but I'll leave the comment here for documentation purposes)

I don't think we need to make bindings for utility classes like Dict that aren't part of the core computation graph library. You can use IndexedList from jklol to get essentially the same interface.

PS the example trains an LSTM to predict the length of a sentence

joelgrus commented:
ok, I removed all of the `dict` bindings. Rather than have dynet take a dependency on jklol, I just implemented what I needed in the example. if I need it for other examples I'll pull it out.
